pkg/services/thumbs/crawler_auth.go

```diff
@@ -2,40 +2,137 @@ package thumbs
+func ProvideCrawlerAuthSetupService(serviceAccounts serviceaccounts.Service, serviceAccountsStore serviceaccounts.Store, sqlStore *sqlstore.SQLStore) *OSSCrawlerAuthSetupService {
+	return &OSSCrawlerAuthSetupService{
+		serviceAccountNamePrefix: "dashboard-previews-crawler-org-",
+		serviceAccounts:          serviceAccounts,
+		log:                      log.New("oss_crawler_account_setup_service"),
+		sqlStore:                 sqlStore,
+		serviceAccountsStore:     serviceAccountsStore,
+	}
 }
 
-type OSSCrawlerAuthSetupService struct{}
+type OSSCrawlerAuthSetupService struct {
+	log                      log.Logger
+	serviceAccountNamePrefix string
+	serviceAccounts          serviceaccounts.Service
+	serviceAccountsStore     serviceaccounts.Store
+	sqlStore                 *sqlstore.SQLStore
+}
 
 type CrawlerAuth interface {
 	GetUserId(orgId int64) int64
+	GetLogin(orgId int64) string
 	GetOrgRole() models.RoleType
 }
 
-type staticCrawlerAuth struct {
-	userId  int64
-	orgRole models.RoleType
+func (o *OSSCrawlerAuthSetupService) findAllOrgIds(ctx context.Context) ([]int64, error) {
+	searchAllOrgsQuery := &models.SearchOrgsQuery{}
+	if err := o.sqlStore.SearchOrgs(ctx, searchAllOrgsQuery); err != nil {
+		o.log.Error("Error when searching for orgs", "err", err)
+		return nil, err
+	}
+
+	orgIds := make([]int64, 0)
+	for i := range searchAllOrgsQuery.Result {
+		orgIds = append(orgIds, searchAllOrgsQuery.Result[i].Id)
+	}
+
+	return orgIds, nil
+}
+
+type crawlerAuth struct {
+	accountIdByOrgId map[int64]int64
+	loginByOrgId     map[int64]string
+	orgRole          models.RoleType
 }
 
-func (o *staticCrawlerAuth) GetOrgRole() models.RoleType {
+func (o *crawlerAuth) GetOrgRole() models.RoleType {
 	return o.orgRole
 }
 
-func (o *staticCrawlerAuth) GetUserId(orgId int64) int64 {
-	return o.userId
+func (o *crawlerAuth) GetUserId(orgId int64) int64 {
+	return o.accountIdByOrgId[orgId]
+}
+
+func (o *crawlerAuth) GetLogin(orgId int64) string {
+	return o.loginByOrgId[orgId]
 }
 
 func (o *OSSCrawlerAuthSetupService) Setup(ctx context.Context) (CrawlerAuth, error) {
+	orgIds, err := o.findAllOrgIds(ctx)
+	if err != nil {
+		return nil, err
+	}
+
 	// userId:0 and ROLE_ADMIN grants the crawler process permissions to view all dashboards in all folders & orgs
 	// the process doesn't and shouldn't actually need to edit/modify any resources from the UI
-	return &staticCrawlerAuth{userId: 0, orgRole: models.ROLE_ADMIN}, nil
+	orgRole := models.ROLE_ADMIN
+
+	accountIdByOrgId := make(map[int64]int64)
+	loginByOrgId := make(map[int64]string)
+	for _, orgId := range orgIds {
+		o.log.Info("Creating account for org", "orgId", orgId)
+
+		serviceAccountNameOrg := o.serviceAccountNamePrefix + strconv.FormatInt(orgId, 10)
+
+		saForm := serviceaccounts.CreateServiceAccountForm{
+			Name: serviceAccountNameOrg,
+			Role: &orgRole,
+		}
+
+		serviceAccount, err := o.serviceAccounts.CreateServiceAccount(context.Background(), orgId, &saForm)
```

Comment:
can `ctx` be passed to this call?

Reply:
good catch, done both here and in the ent